fontbe/src/hvar.rs

```diff
@@ -0,0 +1,233 @@
+                        values[0].ot_round(),
+                    ))
+                })
+                .collect(),
+        );
+        Ok(())
+    }
+
+    fn is_single_model(&self) -> bool {
+        self.models.len() == 1
+    }
+
+    fn iter(&self) -> impl Iterator<Item = &Vec<(VariationRegion, i16)>> {
+        self.deltas.iter()
+    }
+}
+
+impl Work<Context, AnyWorkId, Error> for HvarWork {
+    fn id(&self) -> AnyWorkId {
+        WorkId::Hvar.into()
+    }
+
+    fn read_access(&self) -> Access<AnyWorkId> {
+        Access::Custom(Arc::new(|id| {
+            matches!(
+                id,
+                AnyWorkId::Fe(FeWorkId::Glyph(..))
+                    | AnyWorkId::Fe(FeWorkId::StaticMetadata)
+                    | AnyWorkId::Fe(FeWorkId::GlyphOrder)
+            )
+        }))
+    }
+
+    /// Generate [HVAR](https://learn.microsoft.com/en-us/typography/opentype/spec/HVAR)
+    fn exec(&self, context: &Context) -> Result<(), Error> {
+        let static_metadata = context.ir.static_metadata.get();
+        let var_model = &static_metadata.variation_model;
+        let glyph_order = context.ir.glyph_order.get();
+
+        let mut glyph_width_deltas = AdvanceWidthDeltas::new(var_model.clone());
+        for name in glyph_order.iter() {
+            let glyph = context.ir.glyphs.get(&FeWorkId::Glyph(name.clone()));
+            glyph_width_deltas.add(glyph.as_ref())?;
+        }
+
+        let mut var_idxes = Vec::new();
+
+        // if we have a single model, we can try to build a VariationStore with implicit variation
+        // indices (a single ItemVariationData, outer index 0, inner index => gid).
+        let direct_store = if glyph_width_deltas.is_single_model() {
+            let mut direct_builder = VariationStoreBuilder::new_with_implicit_indices();
+            for deltas in glyph_width_deltas.iter() {
+                var_idxes.push(direct_builder.add_deltas(deltas.clone()));
+            }
+            // sanity check
+            assert_eq!(
+                mem::take(&mut var_idxes),
+                (0..glyph_order.len() as u32).collect::<Vec<_>>()
+            );
+            // we don't use the returned (identity) map in this case
+            Some(direct_builder.build().0)
+        } else {
+            None
+        };
+
+        // also build an indirect VariationStore with a DeltaSetIndexMap to map gid => varidx
+        let mut indirect_builder = VariationStoreBuilder::new();
+        for deltas in glyph_width_deltas.iter() {
+            var_idxes.push(indirect_builder.add_deltas(deltas.clone()));
+        }
+        let (indirect_store, varidx_map) = indirect_builder.build();
+
+        let varidx_map: DeltaSetIndexMap = var_idxes
+            .into_iter()
+            .map(|idx| {
+                varidx_map
+                    .get(idx)
+                    .unwrap_or_else(|| panic!("varidx_map missing index {}", idx))
```

Comment:
I think just `expect` is fine here? it feels strange to use `unwrap_or_else` in this way although I get the motivation (I think? to avoid the allocation of using `format!` in `expect`?)

Reply:
yes, exactly that. This should never trigger basically, if the VariationStoreBuilder holds on to its contract

Reply:
It's an invariant that any temporary index returned by `add_deltas` will exist in the returned map; if this is ever not true then something is seriously wrong, so I would personally just `unwrap` or `expect` here, but this is fine.

Reply:
ok, I'll unwrap() then